docs/posts/llms-and-data/index.qmd

```diff
@@ -0,0 +1,258 @@
+
+## Approaches
+
+When discussed internally at Voltron Data, we identified three distinct
+approaches to applying LLMs to data analytics that can be implemented today:
+
+1. LLM writes SQL
+2. LLM writes a subroutine (Python UDF or another language)
+3. Use LLM in a subroutine
+
+While these three approaches are not an exhaustive list of how LLMs can be
+applied to data, they can be easily understood and implemented with
+Ibis and [Marvin](https://www.askmarvin.ai/welcome/what_is_marvin/), an AI
+engineering framework for building natural language interfaces. Together with
+these two open-source tools, we can build a natural language interface for
+data analytics that supports 18+ backends. But first, let's demonstrate the
+three approaches in code.
+
+### Approach 1: LLM writes SQL
+
+State of the art (SoTA) LLMs are decent at generating SQL out of the box. We can
+be clever to handle errors, retries, and more, but in its simplest form:
+
+```{python}
+# | code-fold: true
+import ibis  # <1>
+import marvin  # <1>
+
+from dotenv import load_dotenv  # <1>
+
+load_dotenv()  # <2>
+
+con = ibis.connect("duckdb://penguins.ddb")  # <3>
+t = ibis.examples.penguins.fetch()  # <3>
+t = con.create_table("penguins", t.to_pyarrow(), overwrite=True)  # <3>
+```
+
+1. Import the libraries we need.
+2. Load the environment variable to setup Marvin to call our OpenAI account.
+3. Setup the demo datain an Ibis backend.
+
+```{python}
+import ibis  # <1>
+import marvin  # <1>
+
+from ibis.expr.schema import Schema  # <1>
+from ibis.expr.types.relations import Table  # <1>
+
+ibis.options.interactive = True  # <2>
+marvin.settings.llm_model = "openai/gpt-4"  # <2>
+
+con = ibis.connect("duckdb://penguins.ddb")  # <3>
+t = con.table("penguins")  # <3>
+```
+
+1. Import Ibis and Marvin.
+2. Configure Ibis (interactive) and Marvin (GPT-4).
+3. Connect to the data and load a table into a variable.
+
+```{python}
+@marvin.ai_fn  # <1>
+def _generate_sql_select(
+    text: str, table_name: str, table_schema: Schema
+) -> str:  # <1>
+    """Generate SQL SELECT from text."""  # <1>
+
+
+def sql_from_text(text: str, t: Table) -> Table:  # <2>
+    """Run SQL from text."""  # <2>
+    return t.sql(_generate_sql_select(text, t.get_name(), t.schema()).strip(";"))  # <2>
+```
+
+1. A non-deterministic, LLM-powered AI function.
+2. A deterministic, human-authored function that calls the AI function.
+
+```{python}
+t2 = sql_from_text("the unique combination of species and islands", t)
+t2
+```
+
+```{python}
+t3 = sql_from_text(
+    "the unique combination of species and islands, with their counts, ordered from highest to lowest, and name that column just 'count'",
+    t,
+)
+t3
+```
+
+This works well-enough for simple cases and can be expanded to handle complex
+ones. In many scenarios, it may be easier to express a query in English or
+another language than to write it in SQL, especially if working across multiple
+SQL dialects.
+
```

Comment:
What about Ibis? Can we show Birdbrain writing Ibis code here? Or can we briefly explain that it _could_ also write Ibis code or code for other dataframe libraries?